hydrolib/core/io/bc/models.py

```diff
@@ -136,50 +50,107 @@ def _supports_comments(cls):
     def _duplicate_keys_as_list(cls):
         return True
 
-    _header: Literal["Forcing"] = "Forcing"
+    def _convert_section_to_dict(self) -> Dict:
+        return self.dict(
+            exclude={
+                "start_line",
+                "end_line",
+            }
+        )
 
-    name: str
+    @classmethod
+    def validate(cls, v):
+        """Try to iniatialize subclass based on function field."""
+        # should be replaced by discriminated unions once merged
+        # https://github.com/samuelcolvin/pydantic/pull/2336
+        if isinstance(v, dict):
+            for c in cls.__subclasses__():
+                if (
+                    c.__fields__.get("function").default
+                    == v.get("function", "").lower()
+                ):
+                    v = c(**v)
+                    break
+        return v
+
+
+class TimeSeries(ForcingBase):
+    function: Literal["timeseries"] = "timeseries"
+    timeinterpolation: TimeInterpolation = Field(alias="timeInterpolation")
+    offset: float = 0.0
+    factor: float = 1.0
+
+
+class Harmonic(ForcingBase):
+    function: Literal["harmonic"] = "harmonic"
+    factor: float = 1.0
+
+
+class Astronomic(ForcingBase):
+    function: Literal["astronomic"] = "astronomic"
+    factor: float = 1.0
+
+
+class HarmonicCorrection(ForcingBase):
+    function: Literal["harmoniccorrection"] = "harmoniccorrection"
 
-    function: Union[
-        Function.TimeSeries,
-        Function.Harmonic,
-        Function.Astronomic,
-        Function.HarmonicCorrection,
-        Function.AstronomicCorrection,
-        Function.T3D,
-        Function.QHTable,
-        Function.Constant,
-    ]
+
+class AstronomicCorrection(ForcingBase):
+    function: Literal["astronomiccorrection"] = "astronomiccorrection"
+
+
+class T3D(ForcingBase):
+    function: Literal["t3d"] = "t3d"
+
+    offset: float = 0.0
+    factor: float = 1.0
+
+    vertical_positions: List[float] = Field(alias="verticalPositions")
+    vertical_interpolation: VerticalInterpolation = Field(alias="verticalInterpolation")
+    vertical_position_type: VerticalPositionType = Field(alias="verticalPositionType")
+
+
+class QHTable(ForcingBase):
+    function: Literal["qhtable"] = "qhtable"
+
+
+class Constant(ForcingBase):
```

Comment:
Technically a constant has only a single value. It might be nice to add a follow up issue to expose it as such.

Reply:
Well, we might need helper functions for all these classes, although I wouldn't be surprised if this class can represent multiple static values (instead of a series).